### PR TITLE
Restore Fusion index to that of 3.3

### PR DESCRIPTION
**Pull request by burqen**
Disable
- Native string index
- Native Temporal index
- Native Spatial index

Also make Fusion index multiversionable for future adaptation.

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/helpers/ArrayUtil.java [hunk 1]
```diff
@@ -431,14 +433,33 @@ public static boolean isEmpty( Object[] array )
...
-            if ( array[i].equals( item ) )
+            if ( item.equals( array[i] ) )
             {
                 return i;
             }
         }
         return -1;
     }
 
+    public static int[] without( int[] source, int... toRemove )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/helpers/ArrayUtil.java]: please use ArrayUtils.removeAll instead

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexAccessor.java [hunk 2]
```diff
@@ -100,7 +102,10 @@ public IndexReader newReader()
     @Override
     public BoundedIterable<Long> newAllEntriesReader()
     {
-        BoundedIterable<Long>[] entries = instancesAs( BoundedIterable.class, IndexAccessor::newAllEntriesReader );
+        @SuppressWarnings( "unchecked" )
+        BoundedIterable<Long>[] entries = Arrays.stream( instancesAs( BoundedIterable.class, IndexAccessor::newAllEntriesReader ) )
+                .filter( notNull() )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexAccessor.java]: i think filtering out and guarding against nulls everywhere is a source of errors, what if instead we will have those instance as an empty instance of some sort?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSelector00.java [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+ */
+package org.neo4j.kernel.impl.index.schema.fusion;
+
+import org.neo4j.internal.kernel.api.IndexQuery;
+import org.neo4j.storageengine.api.schema.IndexReader;
+import org.neo4j.values.storable.Value;
+
+import static org.neo4j.kernel.impl.index.schema.fusion.FusionIndexBase.LUCENE;
+
+public class FusionSelector00 implements FusionIndexProvider.Selector
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSelector00.java]: java docs are not here at all right now (in all of the selectors)

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]
Couple of macro comments, lets talk about those

**Comment by burqen** [comment]
@MishaDemianenko Sorry for force pushing the fixes, but it was the easiest way to revert all of the null checks. This looks a lot better I think. Thanks for the feedback!

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexBase.java [hunk 4]
```diff
@@ -163,4 +166,19 @@
             throw exception;
         }
     }
+
+    static void validateSelectorInstances( Object[] instances, int... notNullIndex )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexBase.java]: thats not really notNullIndex now

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java [hunk 5]
```diff
@@ -76,103 +71,28 @@ public IndexSampler createSampler()
     @Override
     public PrimitiveLongResourceIterator query( IndexQuery... predicates ) throws IndexNotApplicableKernelException
     {
-        if ( predicates.length > 1 )
+        IndexReader instance = selector.select( instances, predicates );
+        if ( instance != null )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:75]: thats no longer so right?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java [hunk 6]
```diff
@@ -76,103 +71,28 @@ public IndexSampler createSampler()
...
-             predicate instanceof StringSuffixPredicate ||
-             predicate instanceof StringContainsPredicate )
-        {
-            instances[STRING].query( cursor, indexOrder, predicate );
-            return;
-        }
-
-        if ( predicate instanceof RangePredicate )
+        IndexReader instance = selector.select( instances, predicates );
+        if ( instance != null )
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionIndexReader.java:91]: and here?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSelector00.java [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+import org.neo4j.internal.kernel.api.IndexQuery;
+import org.neo4j.storageengine.api.schema.IndexReader;
+import org.neo4j.values.storable.Value;
+
+import static org.neo4j.kernel.impl.index.schema.fusion.FusionIndexBase.LUCENE;
+
+/**
+ * Selector for index provider "lucene-1.x".
+ * The version name "00" comes from lucene-1.x originally not being a fusion index.
+ * todo If spatial and temporal will live together with lucene-1.x, update this selector. Otherwise, remove it.
```
Comment (MishaDemianenko) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/index/schema/fusion/FusionSelector00.java:31]: temporal

**Review by MishaDemianenko — CHANGES_REQUESTED** [review]
Thats indeed much better but still have some questions

**Review by MishaDemianenko — APPROVED** [review]

**Comment by tinwelint** [comment]
Closing and reopening when development has gotten further.

**Pull request closed without merging**